File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/globalcontexts/GlobalContextTest.java
package com.snowplowanalytics.snowplow.globalcontexts;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;

import com.snowplowanalytics.snowplow.Snowplow;
import com.snowplowanalytics.snowplow.configuration.GlobalContextsConfiguration;
import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration;
import com.snowplowanalytics.snowplow.configuration.TrackerConfiguration;
import com.snowplowanalytics.snowplow.controller.TrackerController;
import com.snowplowanalytics.snowplow.event.AbstractPrimitive;
import com.snowplowanalytics.snowplow.event.AbstractSelfDescribing;
import com.snowplowanalytics.snowplow.event.ScreenView;
import com.snowplowanalytics.snowplow.event.Structured;
import com.snowplowanalytics.snowplow.event.Timing;
import com.snowplowanalytics.snowplow.internal.constants.Parameters;
import com.snowplowanalytics.snowplow.tracker.InspectableEvent;
import com.snowplowanalytics.snowplow.internal.tracker.ServiceProvider;
import com.snowplowanalytics.snowplow.internal.tracker.TrackerEvent;
import com.snowplowanalytics.snowplow.network.HttpMethod;
import com.snowplowanalytics.snowplow.network.Protocol;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

class GlobalContextGenerator implements ContextGenerator {

    @Override
    public boolean filterEvent(@NonNull InspectableEvent event) {
        return "StringToMatch".equals(event.getPayload().get(Parameters.SE_CATEGORY));
    }

    @NonNull
    @Override
    public List<SelfDescribingJson> generateContexts(@NonNull InspectableEvent event) {
        return Collections.singletonList(new SelfDescribingJson("schema", new HashMap<String, String>() {{
            put("key", "value");
        }}));
    }
}


@RunWith(AndroidJUnit4.class)
public class GlobalContextTest {

    @Test
    public void testGlobalContexts() {
        SelfDescribingJson sdj = new SelfDescribingJson("schema", new HashMap<String, String>() {{ put("key", "value"); }});
        GlobalContext staticGC = new GlobalContext(Collections.singletonList(sdj));
        GlobalContext generatorGC = new GlobalContext(new GlobalContextGenerator());
        GlobalContext blockGC = new GlobalContext(new FunctionalGenerator() {
            @Nullable
            @Override
            public List<SelfDescribingJson> apply(@NonNull InspectableEvent event) {
                return Collections.singletonList(
                        new SelfDescribingJson("schemaBlock", new HashMap<String, String>() {{ put("key", "value"); }})
                );
            }
        });
        TrackerController tracker = getTracker(new HashMap<String,GlobalContext>() {{
            put("static", staticGC);
            put("generator", generatorGC);
            put("block", blockGC);
        }});

        Set<String> result = tracker.getGlobalContexts().getTags();
        Set<String> expected = Set.of("static", "generator", "block");
        assertEquals(expected, result);

        // Can't remove a not existing tag
        GlobalContext removedGC = tracker.getGlobalContexts().remove("notExistingTag");
        assertNull(removedGC);
        result = tracker.getGlobalContexts().getTags();
        expected = Set.of("static","generator","block");
        assertEquals(expected, result);

        // Remove an existing tag
        removedGC = tracker.getGlobalContexts().remove("static");
        assertNotNull(removedGC);
        result = tracker.getGlobalContexts().getTags();
        expected = Set.of("generator","block");
        assertEquals(expected, result);

        // Add a not existing tag
        assertTrue(tracker.getGlobalContexts().add("static", staticGC));
        result = tracker.getGlobalContexts().getTags();
        expected = Set.of("static","generator","block");
        assertEquals(expected, result);

        // Can't add an existing tag
        assertFalse(tracker.getGlobalContexts().add("static", staticGC));
        result = tracker.getGlobalContexts().getTags();
        expected = Set.of("static","generator","block");
        assertEquals(expected, result);
    }

    @Test
    public void testAddRemoveGlobalContexts() {
        SelfDescribingJson sdj = new SelfDescribingJson("schema", new HashMap<String, String>() {{ put("key", "value"); }});
        GlobalContext staticGC = new GlobalContext(Collections.singletonList(sdj));
        TrackerController tracker = getTracker(null);

        Set<String> result = tracker.getGlobalContexts().getTags();
        Set<String> expected = Set.of();
        assertEquals(expected, result);

        // Can't remove a not existing tag
        GlobalContext removedGC = tracker.getGlobalContexts().remove("notExistingTag");
        assertNull(removedGC);

        // Add a not existing tag
        assertTrue(tracker.getGlobalContexts().add("static", staticGC));
        result = tracker.getGlobalContexts().getTags();
        expected = Set.of("static");
        assertEquals(expected, result);

        // Remove an existing tag
        removedGC = tracker.getGlobalContexts().remove("static");
        assertNotNull(removedGC);
        result = tracker.getGlobalContexts().getTags();
        expected = Set.of();
        assertEquals(expected, result);
    }

    @Test
    public void testStaticGenerator() {
        SelfDescribingJson sdj = new SelfDescribingJson("schema", new HashMap<String, String>() {{ put("key", "value"); }});
        GlobalContext staticGC = new GlobalContext(Collections.singletonList(sdj));

        AbstractPrimitive event = new Structured("Category", "Action");
        TrackerEvent trackerEvent = new TrackerEvent(event);

        List<SelfDescribingJson> contexts = staticGC.generateContexts(trackerEvent);
        assertEquals(1, contexts.size());
        assertEquals("schema", contexts.get(0).getMap().get("schema"));
    }

    @Test
    public void testStaticGeneratorWithFilter() {
        String stringToMatch = "StringToMatch";
        SelfDescribingJson sdj = new SelfDescribingJson("schema", new HashMap<String, String>() {{ put("key", "value"); }});
        GlobalContext filterMatchingGC = new GlobalContext(Collections.singletonList(sdj), new FunctionalFilter() {
            @Override
            public boolean apply(@NonNull InspectableEvent event) {
                return stringToMatch.equals(event.getPayload().get(Parameters.SE_CATEGORY));
            }
        });

        AbstractPrimitive event = new Structured(stringToMatch, "Action");
        TrackerEvent trackerEvent = new TrackerEvent(event);

        List<SelfDescribingJson> contexts = filterMatchingGC.generateContexts(trackerEvent);
        assertEquals(1, contexts.size());
        assertEquals("schema", contexts.get(0).getMap().get("schema"));

        // Not Matching
        GlobalContext filterNotMatchingGC = new GlobalContext(Collections.singletonList(sdj), new FunctionalFilter() {
            @Override
            public boolean apply(@NonNull InspectableEvent event) {
                return false;
            }
        });

        contexts = filterNotMatchingGC.generateContexts(trackerEvent);
        assertEquals(0, contexts.size());
    }

    @Test
    public void testStaticGeneratorWithRuleSet() {
        String allowed = "iglu:com.snowplowanalytics.*/*/jsonschema/*-*-*";
        String denied = "iglu:com.snowplowanalytics.mobile/*/jsonschema/*-*-*";
        SchemaRuleSet ruleset = SchemaRuleSet.buildRuleSet(List.of(allowed), List.of(denied));

        SelfDescribingJson sdj = new SelfDescribingJson("schema", new HashMap<String, String>() {{ put("key", "value"); }});
        GlobalContext rulesetGC = new GlobalContext(Collections.singletonList(sdj), ruleset);

        // Not matching primitive event
        AbstractPrimitive event = new Structured("Category", "Action");
        TrackerEvent trackerEvent = new TrackerEvent(event);
        List<SelfDescribingJson> contexts = rulesetGC.generateContexts(trackerEvent);
        assertEquals(0, contexts.size());

        // Not matching self-describing event with mobile schema
        AbstractSelfDescribing selfDescribingEvent = new ScreenView("Name", null)
                .type("Type");
        trackerEvent = new TrackerEvent(selfDescribingEvent);
        contexts = rulesetGC.generateContexts(trackerEvent);
        assertEquals(0, contexts.size());

        // Matching self-describing event with general schema
        selfDescribingEvent = new Timing("Category", "Variable", 123)
                .label("Label");
        trackerEvent = new TrackerEvent(selfDescribingEvent);
        contexts = rulesetGC.generateContexts(trackerEvent);
        assertEquals(1, contexts.size());
        assertEquals("schema", contexts.get(0).getMap().get("schema"));
    }

    @Test
    public void testBlockGenerator() {
        GlobalContext blockGC = new GlobalContext(new FunctionalGenerator() {
            @Nullable
            @Override
            public List<SelfDescribingJson> apply(@NonNull InspectableEvent event) {
                return Collections.singletonList(
                        new SelfDescribingJson("schemaBlock", new HashMap<String, String>() {{ put("key", "value"); }})
                );
            }
        });

        AbstractPrimitive event = new Structured("Category", "Action");
        TrackerEvent trackerEvent = new TrackerEvent(event);

        List<SelfDescribingJson> contexts = blockGC.generateContexts(trackerEvent);
        assertEquals(1, contexts.size());
        assertEquals("schemaBlock", contexts.get(0).getMap().get("schema"));
    }

    @Test
    public void testContextGenerator() {
        GlobalContext contextGeneratorGC = new GlobalContext(new GlobalContextGenerator());
        AbstractPrimitive event = new Structured("StringToMatch", "Action");
        TrackerEvent trackerEvent = new TrackerEvent(event);

        List<SelfDescribingJson> contexts = contextGeneratorGC.generateContexts(trackerEvent);
        assertEquals(1, contexts.size());
        assertEquals("schema", contexts.get(0).getMap().get("schema"));
    }

    // Service methods

    private TrackerController getTracker(Map<String, GlobalContext> generators) {
        Context context = InstrumentationRegistry.getInstrumentation().getContext();
        NetworkConfiguration networkConfig = new NetworkConfiguration("com.acme.fake", HttpMethod.POST);
        TrackerConfiguration trackerConfig = new TrackerConfiguration("anAppId")
                .platformContext(true)
                .geoLocationContext(false)
                .base64encoding(false)
                .sessionContext(true);
        GlobalContextsConfiguration gcConfig = new GlobalContextsConfiguration(generators);
        return Snowplow.createTracker(context, "aNamespace", networkConfig, trackerConfig, gcConfig);
    }
}

File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/globalcontexts/GlobalContextTest.kt
/*
 * Copyright (c) 2015-present Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.globalcontexts

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.snowplowanalytics.core.constants.Parameters
import com.snowplowanalytics.core.tracker.TrackerEvent
import com.snowplowanalytics.snowplow.Snowplow.createTracker
import com.snowplowanalytics.snowplow.configuration.GlobalContextsConfiguration
import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration
import com.snowplowanalytics.snowplow.configuration.TrackerConfiguration
import com.snowplowanalytics.snowplow.controller.TrackerController
import com.snowplowanalytics.snowplow.event.*
import com.snowplowanalytics.snowplow.globalcontexts.SchemaRuleSet.Companion.buildRuleSet
import com.snowplowanalytics.snowplow.network.HttpMethod
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.tracker.InspectableEvent
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

internal class GlobalContextGenerator : ContextGenerator {
    override fun filterEvent(event: InspectableEvent): Boolean {
        return "StringToMatch" == event.payload[Parameters.SE_CATEGORY]
    }

    override fun generateContexts(event: InspectableEvent): List<SelfDescribingJson> {
        return listOf(SelfDescribingJson("schema", object : HashMap<String?, String?>() {
            init {
                put("key", "value")
            }
        }))
    }
}

@RunWith(AndroidJUnit4::class)
class GlobalContextTest {
    @Test
    fun testGlobalContexts() {
        val sdj = SelfDescribingJson("schema", object : HashMap<String?, String?>() {
            init {
                put("key", "value")
            }
        })
        val staticGC = GlobalContext(listOf(sdj))
        val generatorGC = GlobalContext(GlobalContextGenerator())
        val blockGC = GlobalContext(object : FunctionalGenerator() {
            override fun apply(event: InspectableEvent): List<SelfDescribingJson> {
                return listOf(
                    SelfDescribingJson("schemaBlock", object : HashMap<String?, String?>() {
                        init {
                            put("key", "value")
                        }
                    })
                )
            }
        })
        val tracker = getTracker(object : HashMap<String, GlobalContext>() {
            init {
                put("static", staticGC)
                put("generator", generatorGC)
                put("block", blockGC)
            }
        })
        var result = tracker.globalContexts.tags
        var expected = mutableSetOf("static", "generator", "block")
        Assert.assertEquals(expected, result)

        // Can't remove a not existing tag
        var removedGC = tracker.globalContexts.remove("notExistingTag")
        Assert.assertNull(removedGC)
        result = tracker.globalContexts.tags
        expected = mutableSetOf("static", "generator", "block")
        Assert.assertEquals(expected, result)

        // Remove an existing tag
        removedGC = tracker.globalContexts.remove("static")
        Assert.assertNotNull(removedGC)
        result = tracker.globalContexts.tags
        expected = mutableSetOf("generator", "block")
        Assert.assertEquals(expected, result)

        // Add a not existing tag
        Assert.assertTrue(tracker.globalContexts.add("static", staticGC))
        result = tracker.globalContexts.tags
        expected = mutableSetOf("static", "generator", "block")
        Assert.assertEquals(expected, result)

        // Can't add an existing tag
        Assert.assertFalse(tracker.globalContexts.add("static", staticGC))
        result = tracker.globalContexts.tags
        expected = mutableSetOf("static", "generator", "block")
        Assert.assertEquals(expected, result)
    }

    @Test
    fun testAddRemoveGlobalContexts() {
        val sdj = SelfDescribingJson("schema", object : HashMap<String?, String?>() {
            init {
                put("key", "value")
            }
        })
        val staticGC = GlobalContext(listOf(sdj))
        val tracker = getTracker(null)
        var result = tracker.globalContexts.tags
        var expected = mutableSetOf<String?>()
        Assert.assertEquals(expected, result)

        // Can't remove a not existing tag
        var removedGC = tracker.globalContexts.remove("notExistingTag")
        Assert.assertNull(removedGC)

        // Add a not existing tag
        Assert.assertTrue(tracker.globalContexts.add("static", staticGC))
        result = tracker.globalContexts.tags
        expected = mutableSetOf("static")
        Assert.assertEquals(expected, result)

        // Remove an existing tag
        removedGC = tracker.globalContexts.remove("static")
        Assert.assertNotNull(removedGC)
        result = tracker.globalContexts.tags
        expected = mutableSetOf()
        Assert.assertEquals(expected, result)
    }

    @Test
    fun testStaticGenerator() {
        val sdj = SelfDescribingJson("schema", object : HashMap<String?, String?>() {
            init {
                put("key", "value")
            }
        })
        val staticGC = GlobalContext(listOf(sdj))
        val event: AbstractPrimitive = Structured("Category", "Action")
        val trackerEvent = TrackerEvent(event)
        val contexts = staticGC.generateContexts(trackerEvent)
        Assert.assertEquals(1, contexts.size.toLong())
        Assert.assertEquals("schema", contexts[0].map["schema"])
    }

    @Test
    fun testStaticGeneratorWithFilter() {
        val stringToMatch = "StringToMatch"
        val sdj = SelfDescribingJson("schema", object : HashMap<String?, String?>() {
            init {
                put("key", "value")
            }
        })
        val filterMatchingGC = GlobalContext(listOf(sdj), object : FunctionalFilter() {
            override fun apply(event: InspectableEvent): Boolean {
                return stringToMatch == event.payload[Parameters.SE_CATEGORY]
            }
        })
        val event: AbstractPrimitive = Structured(stringToMatch, "Action")
        val trackerEvent = TrackerEvent(event)
        var contexts = filterMatchingGC.generateContexts(trackerEvent)
        Assert.assertEquals(1, contexts.size.toLong())
        Assert.assertEquals("schema", contexts[0].map["schema"])

        // Not Matching
        val filterNotMatchingGC = GlobalContext(listOf(sdj), object : FunctionalFilter() {
            override fun apply(event: InspectableEvent): Boolean {
                return false
            }
        })
        contexts = filterNotMatchingGC.generateContexts(trackerEvent)
        Assert.assertEquals(0, contexts.size.toLong())
    }

    @Test
    fun testStaticGeneratorWithRuleSet() {
        val allowed = "iglu:com.snowplowanalytics.*/*/jsonschema/*-*-*"
        val denied = "iglu:com.snowplowanalytics.mobile/*/jsonschema/*-*-*"
        val ruleset = buildRuleSet(listOf(allowed), listOf(denied))
        val sdj = SelfDescribingJson("schema", object : HashMap<String?, String?>() {
            init {
                put("key", "value")
            }
        })
        val rulesetGC = GlobalContext(listOf(sdj), ruleset)

        // Not matching primitive event
        val event: AbstractPrimitive = Structured("Category", "Action")
        var trackerEvent = TrackerEvent(event)
        var contexts = rulesetGC.generateContexts(trackerEvent)
        Assert.assertEquals(0, contexts.size.toLong())

        // Not matching self-describing event with mobile schema
        var selfDescribingEvent: AbstractSelfDescribing = ScreenView("Name", null)
            .type("Type")
        trackerEvent = TrackerEvent(selfDescribingEvent)
        contexts = rulesetGC.generateContexts(trackerEvent)
        Assert.assertEquals(0, contexts.size.toLong())

        // Matching self-describing event with general schema
        selfDescribingEvent = Timing("Category", "Variable", 123)
            .label("Label")
        trackerEvent = TrackerEvent(selfDescribingEvent)
        contexts = rulesetGC.generateContexts(trackerEvent)
        Assert.assertEquals(1, contexts.size.toLong())
        Assert.assertEquals("schema", contexts[0].map["schema"])
    }

    @Test
    fun testBlockGenerator() {
        val blockGC = GlobalContext(object : FunctionalGenerator() {
            override fun apply(event: InspectableEvent): List<SelfDescribingJson> {
                return listOf(
                    SelfDescribingJson("schemaBlock", object : HashMap<String?, String?>() {
                        init {
                            put("key", "value")
                        }
                    })
                )
            }
        })
        val event: AbstractPrimitive = Structured("Category", "Action")
        val trackerEvent = TrackerEvent(event)
        val contexts = blockGC.generateContexts(trackerEvent)
        Assert.assertEquals(1, contexts.size.toLong())
        Assert.assertEquals("schemaBlock", contexts[0].map["schema"])
    }

    @Test
    fun testContextGenerator() {
        val contextGeneratorGC = GlobalContext(GlobalContextGenerator())
        val event: AbstractPrimitive = Structured("StringToMatch", "Action")
        val trackerEvent = TrackerEvent(event)
        val contexts = contextGeneratorGC.generateContexts(trackerEvent)
        Assert.assertEquals(1, contexts.size.toLong())
        Assert.assertEquals("schema", contexts[0].map["schema"])
    }

    // Service methods
    private fun getTracker(generators: MutableMap<String, GlobalContext>?): TrackerController {
        val context = InstrumentationRegistry.getInstrumentation().context
        val networkConfig = NetworkConfiguration("com.acme.fake", HttpMethod.POST)
        val trackerConfig = TrackerConfiguration("anAppId")
            .platformContext(true)
            .geoLocationContext(false)
            .base64encoding(false)
            .sessionContext(true)
        val gcConfig = GlobalContextsConfiguration(generators)
        return createTracker(
            context,
            "aNamespace" + Math.random().toString(),
            networkConfig,
            trackerConfig,
            gcConfig
        )
    }
}
